### ISSUE TITLE
[START HERE] GPT-2 in C learning roadmap and AI tutor contract

### ISSUE THREAD

**Issue by 25d-228** [25d-228/gpt-2#1]
## Purpose
Build GPT-2 yourself in small, testable C lessons: understand the forward pass, derive and implement backpropagation, train a tiny model, run inference, and finally verify compatibility with pretrained GPT-2. This is an educational implementation, not a promise to reproduce large-scale pretraining on a laptop.

This issue is the course index, not a one-hour coding assignment. The numbered S001–S096 lesson issues are the assignments. Their linked index will be added here after creation. Follow lesson IDs, not GitHub's newest-first listing.

## Defaults—not prerequisites you must already know
C11; single-threaded CPU; float32 model tensors; row-major contiguous arrays; GCC or Clang and Make; a POSIX shell (Linux/macOS or a suitable Windows environment). Start with tiny synthetic token IDs. No GPU, BLAS, autograd engine, or ML library in the C runtime. Python is allowed only for independent reference fixtures and converting external assets; downloaded weights and Python packages are not needed for the default offline C tests. The exact text-tokenizer extension uses a Unicode-capable C regex library rather than pretending ASCII splitting is GPT-2 tokenization.

The local tutor should discover OS/compiler/RAM and adapt setup in S001. Do not make the learner answer a long questionnaire or choose optimization technologies now.

## Mandatory AI tutor behavior—read before every lesson
1. Read this issue, the selected lesson, its prerequisites, and the relevant repository code. Check actual files and test results; never infer completion from a closed issue alone.
2. Teach, do not take over. Explain the new C concept and the tensor/mathematical idea with a small worked example. Ask the learner to predict one result. Offer hints or pseudocode before a complete implementation; do not write the learner's solution unless explicitly requested.
3. Work on one lesson only. Each lesson targets roughly 30–50 minutes INCLUDING explanation, coding, and tests. This is a planning estimate, not a guarantee. At about 40 minutes, stop adding scope; at 60 minutes stop the session. Record the smallest unfinished acceptance criterion as a separately timeboxed continuation. Do not mark unfinished work complete or cram multiple lessons together. Environment/package/download troubleshooting is its own continuation if needed.
4. When the learner says they are ready to test, explain exactly which test file to open/create, the fixture inputs, how expected values were obtained independently, the command to run from the repository root, the expected exit status, and what a failure means. Let the learner execute the command. A claim that code 'looks right' is not test evidence.
5. Use the issue-specific tests, the existing regression suite, and sanitizers. Deliberately break one relevant line temporarily and demonstrate that the new test fails; restore it. Never weaken tolerances or regenerate expected outputs merely to make a failing implementation pass.
6. Close only after the learner approves, all required tests actually pass, and the learner explains the key idea in their own words. Keep a short `docs/learning-log.md` entry: issue ID, changed files, exact commands/results, one misconception resolved, and any continuation. Never fabricate logs or elapsed time.

## Shared implementation contract
Symbols: `B` batch size, `T` active sequence length, `P` maximum context, `V` vocabulary, `C` hidden width, `L` layers, `H` heads, `D=C/H`. Canonical tiny fixture: `B=2,T=3,P=8,V=16,C=8,L=2,H=2`. GPT-2-small compatibility configuration: `P=1024,V=50257,C=768,L=12,H=12`.

Activations are `[B,T,C]`. Every linear weight is stored `[out,in]`: `y[n,o]=bias[o]+sum_i x[n,i]*W[o,i]`. Packed QKV is `[B,T,3*C]` in Q-then-K-then-V order; split heads are `[B,H,T,D]`; attention matrices are `[B,H,T,T]`. GPT-2 uses learned absolute positions, pre-LayerNorm blocks, LayerNorm epsilon `1e-5`, tanh-approximate GELU, MLP width `4*C`, linear biases, causal self-attention, final LayerNorm, and a tied token embedding/output matrix with NO separate output bias. Dropout is disabled until its dedicated lessons and in every deterministic oracle/gradient check.

Forward functions overwrite outputs. Backward functions ADD into destination gradients (`+=`); the caller zeros gradients at an explicit boundary. Distinct forward input/output buffers unless a lesson explicitly permits aliasing. Shape products and allocation sizes must be overflow-checked. Validate external inputs before entering low-level kernels. Small kernels may require validated positive dimensions. No silent clipping of invalid token IDs or dimensions.

Use `src/` and `include/` for reusable code, `app/` for executable entry points, `tests/test_*.c` for standalone tests, `tests/fixtures/` for small checked-in fixtures, `tools/` for reference/conversion helpers, and ignored `build/`, `data/`, `checkpoints/`, `assets/` for generated/large files. Do not implement an unnecessary general-purpose tensor library.

## Test contract established in S002–S004
`make test TEST=name` builds and runs `tests/test_name.c`; a missing named test MUST fail. `make test` runs all default, offline tests; zero discovered tests MUST fail. `make sanitize TEST=name` rebuilds separately with AddressSanitizer and UndefinedBehaviorSanitizer; `make sanitize` runs the default suite. Tests report the failing index/input, actual and expected values, and return nonzero on failure. Test-only intentional failure flags must not affect ordinary runs.

For ordinary floats, use `abs(actual-expected) <= atol + rtol*abs(expected)` with default `atol=1e-5, rtol=1e-4`; compare IDs/counts/masks exactly. Reject unexpected NaN/Inf explicitly. Numerical gradient checks use a scalar objective, nonuniform upstream gradients, central differences, and a step-size sweep around `1e-3` for float32; start with `atol=2e-3, rtol=2e-2`, report worst error, and investigate instability. Exact/symbolic and independent-reference checks supplement finite differences, especially for tiny gradients. No network, full checkpoint, performance threshold, or random flaky pass criterion in default tests.

Optional heavy tests must be explicitly requested and report a real skip when assets are absent, never a false pass. Compare logits/losses/gradients, not just plausible-looking generated text.

## Progress gates
A: Foundations S001–S010. B: Forward model S011–S030. C: Token-ID inference S031–S034. D: Backpropagation S035–S054. E: Training and persistence S055–S073. F: Pretrained weights and exact text interface S074–S087. G: Optional cache/performance/research extensions S088–S096.

## Getting issues into a local editor
Cloning code does not place issue bodies in the working tree. A tutor with GitHub access can read them directly. With GitHub CLI, run from the clone:
```sh
mkdir -p docs/curriculum
gh issue list --repo 25d-228/gpt-2 --state all --limit 200 --json number,title,body,url > docs/curriculum/issues.json
```
Read the entry for this roadmap and S001; refresh the export when issue requirements change. Do not store credentials or downloaded weights in Git.

## Primary references
Use these to check behavior, not to copy a complete C solution before attempting a lesson. Pin concrete upstream commits/package versions when producing fixtures.
- Original architecture: https://github.com/openai/gpt-2/blob/master/src/model.py
- Original tokenizer: https://github.com/openai/gpt-2/blob/master/src/encoder.py
- GPT-2 configuration/model interface: https://huggingface.co/docs/transformers/model_doc/gpt2
- Released small-model config: https://huggingface.co/openai-community/gpt2/blob/main/config.json
- Independent C comparison, after your own attempt: https://github.com/karpathy/llm.c/blob/master/train_gpt2.c
- Gradient checking: https://docs.pytorch.org/docs/stable/notes/gradcheck.html
- AdamW: https://docs.pytorch.org/docs/stable/generated/torch.optim.AdamW.html
- Sanitizers: https://clang.llvm.org/docs/AddressSanitizer.html and https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html
- Unicode regex: https://www.pcre.org/current/doc/html/pcre2unicode.html
- Issue export: https://cli.github.com/manual/gh_issue_list

## Roadmap acceptance
Keep this issue open as an index. Each lesson needs dependencies, a bounded implementation task, named test files, runnable commands, explicit expected behavior, and a teach-back question. Large-model downloads/runs are opt-in; no production-quality or hardware-speed claims are implied by completion.
